### ISSUE TITLE
Crawler - Non-ASCII filenames cause exception to be thrown

### ISSUE THREAD

**Issue by alxbl** [GoSecure/pyrdp#247]
Found out during live demo :) Callstack to come soon.

**Comment by obilodeau** [comment]
Here is the stacktrace just after successfully saving `secret.txt` from the `C:\`.

Error seems to be in the logging framework not the file IO.

```
--- Logging error ---                                                                                                                                       
Traceback (most recent call last):                                                                                                                             
  File "/usr/lib/python3.6/logging/__init__.py", line 996, in emit                                                                                       
    stream.write(msg)                                                                                                                                      
UnicodeEncodeError: 'ascii' codec can't encode character '\xe9' in position 246: ordinal not in range(128)                                                   
Call stack:                                                                                                                                                 
  File "/opt/venv/bin/pyrdp-mitm.py", line 4, in <module>                                                                                                   
    __import__('pkg_resources').run_script('pyrdp==0.4.2.dev0', 'pyrdp-mitm.py')                                                                            
  File "/opt/venv/lib/python3.6/site-packages/pkg_resources/__init__.py", line 658, in run_script                                                         
    self.require(requires)[0].run_script(script_name, ns)                                                                                                  
  File "/opt/venv/lib/python3.6/site-packages/pkg_resources/__init__.py", line 1438, in run_script                                                               
    exec(code, namespace, namespace)                                                                                                                          
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/EGG-INFO/scripts/pyrdp-mitm.py", line 52, in <module>                
    main()                                                                                                                                               
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/EGG-INFO/scripts/pyrdp-mitm.py", line 45, in main
    main()                                                                                                                                                                                                                                                                        
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/EGG-INFO/scripts/pyrdp-mitm.py", line 45, in main                            
    reactor.run()                                                                                                                                           
  File "/opt/venv/lib/python3.6/site-packages/twisted/internet/asyncioreactor.py", line 267, in run                                                                    
    self._asyncioEventloop.run_forever()                                                                                                                                                                    
  File "/usr/lib/python3.6/asyncio/base_events.py", line 438, in run_forever                                                                             
    self._run_once()                                                                                                                                                  
  File "/usr/lib/python3.6/asyncio/base_events.py", line 1451, in _run_once                                                                                  
    handle._run()                                                                                                                                                                     
  File "/usr/lib/python3.6/asyncio/events.py", line 145, in _run                                                                                             
    self._callback(*self._args)                                                                                                                           
  File "/opt/venv/lib/python3.6/site-packages/twisted/python/log.py", line 103, in callWithLogger                                                            
    return callWithContext({"system": lp}, func, *args, **kw)                                                                                            
  File "/opt/venv/lib/python3.6/site-packages/twisted/python/log.py", line 86, in callWithContext                                                             
    return context.call({ILogContext: newCtx}, func, *args, **kw)                                                                                            
  File "/opt/venv/lib/python3.6/site-packages/twisted/python/context.py", line 122, in callWithContext                                                     
    return self.currentContext().callWithContext(ctx, func, *args, **kw)                                                                                                                         
  File "/opt/venv/lib/python3.6/site-packages/twisted/python/context.py", line 85, in callWithContext                                                      
    return func(*args,**kw)                                                                                                                                  
  File "/opt/venv/lib/python3.6/site-packages/twisted/internet/asyncioreactor.py", line 136, in _readOrWrite                                                       
    why = method()                                                                                                                                     
  File "/opt/venv/lib/python3.6/site-packages/twisted/internet/tcp.py", line 243, in doRead                                                                            
    return self._dataReceived(data)                                                                                                                          
  File "/opt/venv/lib/python3.6/site-packages/twisted/internet/tcp.py", line 249, in _dataReceived                                                       
    rval = self.protocol.dataReceived(data)                                                                                                               
  File "/opt/venv/lib/python3.6/site-packages/twisted/protocols/tls.py", line 330, in dataReceived                                                       
    self._flushReceiveBIO()                                                                                                                               
  File "/opt/venv/lib/python3.6/site-packages/twisted/protocols/tls.py", line 295, in _flushReceiveBIO                                                                                
    ProtocolWrapper.dataReceived(self, bytes)                                                                                                                                                                                                       
  File "/opt/venv/lib/python3.6/site-packages/twisted/protocols/policies.py", line 120, in dataReceived                                                            
    self.wrappedProtocol.dataReceived(data)                                                                                                                 
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/tcp.py", line 91, in dataReceived                                                              
    self.recv(data)                                                                                                                                                                                                                                           
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 143, in recv                                   
    self.pduReceived(pdu)                                                                                                                                                      
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 198, in pduReceived                         
    self.next.recv(pdu.payload)                                                                                                                             
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/segmentation.py", line 79, in recv                          
    layer.recv(forwarded)                                                                                                                                  
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/buffered.py", line 55, in recv                                                       
    self.pduReceived(pdu)                                                                                                                                
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 198, in pduReceived                   
    self.next.recv(pdu.payload)                                                                                                                             
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 143, in recv                                                                                                                                                              
    self.pduReceived(pdu)                                                                                                                                   
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 198, in pduReceived                                                                                                                                           
    self.next.recv(pdu.payload)                                                                                                                                                                                                                                                             
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 143, in recv                                     
    self.pduReceived(pdu)                                                                                                                                   
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 110, in pduReceived                                  
    self.observer.onPDUReceived(pdu)                                                                                                                            
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/core/observer.py", line 82, in __call__                         
    self.composite.doCall(self.item, args, kwargs)                                                                                                             
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/core/observer.py", line 56, in doCall                        
    getattr(observer, item)(*args, **kwargs)                                                                                                               
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 48, in onPDUReceived                       
    self.handlers[pdu.header](pdu)                                                                                                                          
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/MCSMITM.py", line 245, in onSendDataRequest                
    self.clientChannels[pdu.channelID].recv(pdu.payload)                                                                                                    
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mcs/channel.py", line 33, in recv                             
    self.next.recv(data)                                                                                                                                   
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/rdp/security.py", line 142, in recv                            
    self.next.recv(data)                                                                                                                                      
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/rdp/virtual_channel/virtual_channel.py", line 43, in recv
    self.pduReceived(virtualChannelPDU)                                                                                                                  
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 198, in pduReceived
   self.next.recv(pdu.payload)                                                                                                                             
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 143, in recv                                         
    self.pduReceived(pdu)                                                                                                                                                                                   
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 110, in pduReceived                    
    self.observer.onPDUReceived(pdu)                                                                                                                                  
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/core/observer.py", line 82, in __call__                          
    self.composite.doCall(self.item, args, kwargs)                                                                                                                                    
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/core/observer.py", line 56, in doCall                            
    getattr(observer, item)(*args, **kwargs)                                                                                                              
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/DeviceRedirectionMITM.py", line 121, in onClientPDUReceived 
    self.handlePDU(pdu, self.server)                                                                                                                     
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/DeviceRedirectionMITM.py", line 139, in handlePDU            
    self.handleIOResponse(pdu)                                                                                                                               
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/DeviceRedirectionMITM.py", line 170, in handleIOResponse  
    request.handleResponse(pdu)                                                                                                                                                                  
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/DeviceRedirectionMITM.py", line 411, in handleResponse    
    self.handlers[response.majorFunction](response)                                                                                                          
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/DeviceRedirectionMITM.py", line 563, in onDirectoryControlResponse
    self.handleDirectoryListingComplete(response)                                                                                                      
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/DeviceRedirectionMITM.py", line 587, in handleDirectoryListingComplete
    self.mitm.observer.onDirectoryListingComplete(self.deviceID, self.requestID)                                                                             
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/core/observer.py", line 82, in __call__                      
    self.composite.doCall(self.item, args, kwargs)                                                                                                        
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/core/observer.py", line 56, in doCall                        
    getattr(observer, item)(*args, **kwargs)                                                                                                              
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/FileCrawlerMITM.py", line 314, in onDirectoryListingComplete                         
    self.crawlListing(requestID)                                                                                                                                                                                                                    
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/FileCrawlerMITM.py", line 214, in crawlListing                    
    self.fileLogger.info("%(file)s - %(isDirectory)s - %(isDownloaded)s", {"file" : item.path, "isDirectory": item.isDirectory, "isDownloaded": matched})   
Message: '%(file)s - %(isDirectory)s - %(isDownloaded)s'                                                                                                                                         
Arguments: {'file': '/Users/XXX/YYY/Présentation 1.pptx', 'isDirectory': False, 'isDownloaded': True}                                                                                                                               
```

Eventually, it's trying to write the file and then there's an exception and logging the exception fails:

```
--- Logging error ---
Traceback (most recent call last):
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/FileCrawlerMITM.py", line 227, in downloadFile
    targetFile = open(localPath, "wb")
UnicodeEncodeError: 'ascii' codec can't encode character '\xe9' in position 74: ordinal not in range(128)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/logging/__init__.py", line 996, in emit
    stream.write(msg)
UnicodeEncodeError: 'ascii' codec can't encode character '\xe9' in position 189: ordinal not in range(128)
Call stack:
  File "/opt/venv/bin/pyrdp-mitm.py", line 4, in <module>
    __import__('pkg_resources').run_script('pyrdp==0.4.2.dev0', 'pyrdp-mitm.py')
  File "/opt/venv/lib/python3.6/site-packages/pkg_resources/__init__.py", line 658, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/opt/venv/lib/python3.6/site-packages/pkg_resources/__init__.py", line 1438, in run_script
    exec(code, namespace, namespace)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/EGG-INFO/scripts/pyrdp-mitm.py", line 52, in <module>
    main()
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/EGG-INFO/scripts/pyrdp-mitm.py", line 45, in main
    reactor.run()
  File "/opt/venv/lib/python3.6/site-packages/twisted/internet/asyncioreactor.py", line 267, in run
    self._asyncioEventloop.run_forever()
  File "/usr/lib/python3.6/asyncio/base_events.py", line 438, in run_forever
    self._run_once()
  File "/usr/lib/python3.6/asyncio/base_events.py", line 1451, in _run_once
    handle._run()
  File "/usr/lib/python3.6/asyncio/events.py", line 145, in _run
    self._callback(*self._args)
  File "/opt/venv/lib/python3.6/site-packages/twisted/python/log.py", line 103, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "/opt/venv/lib/python3.6/site-packages/twisted/python/log.py", line 86, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/opt/venv/lib/python3.6/site-packages/twisted/python/context.py", line 122, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/opt/venv/lib/python3.6/site-packages/twisted/python/context.py", line 85, in callWithContext
    return func(*args,**kw)
  File "/opt/venv/lib/python3.6/site-packages/twisted/internet/asyncioreactor.py", line 136, in _readOrWrite
    why = method()
  File "/opt/venv/lib/python3.6/site-packages/twisted/internet/tcp.py", line 243, in doRead
    return self._dataReceived(data)
  File "/opt/venv/lib/python3.6/site-packages/twisted/internet/tcp.py", line 249, in _dataReceived
    rval = self.protocol.dataReceived(data)
  File "/opt/venv/lib/python3.6/site-packages/twisted/protocols/tls.py", line 330, in dataReceived
    self._flushReceiveBIO()
  File "/opt/venv/lib/python3.6/site-packages/twisted/protocols/tls.py", line 295, in _flushReceiveBIO
    ProtocolWrapper.dataReceived(self, bytes)
  File "/opt/venv/lib/python3.6/site-packages/twisted/protocols/policies.py", line 120, in dataReceived
    self.wrappedProtocol.dataReceived(data)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/tcp.py", line 91, in dataReceived
    self.recv(data)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 143, in recv
    self.pduReceived(pdu)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 198, in pduReceived
    self.next.recv(pdu.payload)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/segmentation.py", line 79, in recv
    layer.recv(forwarded)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/buffered.py", line 55, in recv
    self.pduReceived(pdu)
[...]
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 110, in pduReceived
    self.observer.onPDUReceived(pdu)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/core/observer.py", line 82, in __call__
    self.composite.doCall(self.item, args, kwargs)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/core/observer.py", line 56, in doCall
    getattr(observer, item)(*args, **kwargs)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/DeviceRedirectionMITM.py", line 121, in onClientPDUReceived
    self.handlePDU(pdu, self.server)                         
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/DeviceRedirectionMITM.py", line 139, in handlePDU
    self.handleIOResponse(pdu)                                   
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/DeviceRedirectionMITM.py", line 170, in handleIOResponse
    request.handleResponse(pdu)                                         
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/DeviceRedirectionMITM.py", line 411, in handleResponse
    self.handlers[response.majorFunction](response)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/DeviceRedirectionMITM.py", line 481, in onReadResponse
    self.handleFileComplete(0)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/DeviceRedirectionMITM.py", line 490, in handleFileComplete
    self.mitm.observer.onFileDownloadComplete(self.deviceID, self.requestID, self.path, error)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/core/observer.py", line 82, in __call__
    self.composite.doCall(self.item, args, kwargs)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/core/observer.py", line 56, in doCall
    getattr(observer, item)(*args, **kwargs)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/FileCrawlerMITM.py", line 293, in onFileDownloadComplete
    self.dispatchDownload()                  
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/FileCrawlerMITM.py", line 149, in dispatchDownload
    self.downloadFile(file)                
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/FileCrawlerMITM.py", line 230, in downloadFile
    self.log.error("Cannot save file: %(localPath)s", {"localPath": localPath})
Message: 'Cannot save file: %(localPath)s'                                                                                    
Arguments: {'localPath': 'pyrdp_output/files/Isabel983278/Users/XXX/YYY/Présentation.pptx'}
```

For logging, we can go UTF-8 (if that's a good practice) but for filesystem artifacts we need to keep portability in mind. Let's hope there's a library or codec that can represent valid Windows filesystem values.

**Comment by alxbl** [comment]
Having everything as UTF-8 should be fine:

https://docs.python.org/3/howto/unicode.html#unicode-filenames

**Comment by alxbl** [comment]
I just tested and I'm not able to reproduce this. I think the problem might be that in Docker environments, `LANG` and `LC_TYPE` might be set to `C`, which means Python will try to encode as ASCII. In any case, I'm going to modify the logger configuration so that file loggers explicitly encode as UTF-8.

We might need to update the Dockerfiles.

**Comment by alxbl** [comment]
```PS D:\dev\PyRDP> docker run --rm -it gosecure/pyrdp:latest-slim bash
pyrdp@1455f23afdb3:~$ echo $LANG

pyrdp@1455f23afdb3:~$ echo $LC_TYPE

pyrdp@1455f23afdb3:~$ python
Python 3.6.9 (default, Jul 17 2020, 12:50:27)
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.getfilesystemencoding()
'ascii'
>>>```

Yep, that's likely the problem.

**Comment by obilodeau** [comment]
Great catch! I got bitten by this `LANG=C` in docker which confuses Python before.

I recalled an article about a fix for Python 3.7: https://lwn.net/Articles/741176/.

Still, we should modify our Dockerfiles or our code. Not sure which is better but I trust your judgment.
